### ISSUE TITLE
Add support for running openHAB with Java 11

### ISSUE THREAD

**Issue by wborn** [openhab/openhab-distro#768]
Currently openHAB only supports Java 8 which is the most recent long-term support (LTS) release of Java. 
We should also add support for running openHAB with Java 11, which is the next LTS release and will be released in a week ([September 25th, 2018](http://openjdk.java.net/projects/jdk/11/)).

In #761 Karaf was upgraded to 4.2.1 which supports Java 8, 9, 10 and 11.

After testing openHAB with Karaf 4.2.1 on Java, 9, 10 and the [JDK 11 Early-Access Builds](http://jdk.java.net/11/) the following issues have been found so far:

* XStream causes the JVM to log warnings for _"Illegal reflective access operations"_
  See: https://github.com/x-stream/xstream/issues/101, https://github.com/eclipse/smarthome/issues/4369#issuecomment-420170606
* RRD4J persistence addon NoSuchMethodError 
  See: https://github.com/rrd4j/rrd4j/issues/72, https://github.com/openhab/openhab1-addons/issues/5663
* MaryTTS voice addon StringIndexOutOfBoundsException 
  See: https://github.com/marytts/marytts/issues/783, https://github.com/openhab/openhab2-addons/issues/3953

For compatibility of Eclipse SmartHome with Java 11 see also [Test ESH with Java 9/10/11](https://github.com/eclipse/smarthome/issues/4369).

---

To simplify JDK 11 testing, I've created an [openhab-docker branch](https://github.com/wborn/openhab-docker/tree/jdk-11) that makes Travis create Debian and Alpine amd64 Docker images using [Zulu Java 11](https://www.azul.com/downloads/zulu/zulu-linux/) every day. 

For the available tags see my [wborn/openhab-jdk11](https://hub.docker.com/r/wborn/openhab-jdk11/tags/) Docker Hub repository.

**Comment by wborn** [comment]
Today Java 11 got released! :gift: :grinning: 

* [Oracle downloads](https://www.oracle.com/technetwork/java/javase/downloads/jdk11-downloads-5066655.html)
* [OpenJDK downloads](http://jdk.java.net/11)

Azul still offers early access (EA) downloads. They do have a [nice Java 11 blog](https://www.azul.com/90-new-features-and-apis-in-jdk-11/).

**Comment by BClark09** [comment]
Debian Buster contains a [armhf package of openJDK 11](https://packages.debian.org/search?keywords=openjdk-11-jdk), so will use that to help test on the RPi if you need! :)

**Comment by wborn** [comment]
> so will use that to help test on the RPi if you need! 

All Java 11 testing is appreciated. :-) Do you know if they compile OpenJDK 11 with hardware floating point operations instead of software floating point operations? AFAIK that usually is the main performance issue with OpenJDK 8 on Raspbian.

**Comment by BClark09** [comment]
I don't know unfortunately. I'm assuming that since the difference between:

```
~$ java --version
openjdk 11 2018-09-25
OpenJDK Runtime Environment (build 11+28-Debian-1)
OpenJDK Server VM (build 11+28-Debian-1, mixed mode)

~$ java -XshowSettings 2>&1 | grep sun.arch
   sun.arch.abi = 
   sun.arch.data.model = 32
```

and 

```
~ $ java -version
openjdk version "1.8.0_152"
OpenJDK Runtime Environment (Zulu Embedded 8.25.0.76-linux-aarch32hf) (build 1.8.0_152-b76)
OpenJDK Client VM (Zulu Embedded 8.25.0.76-linux-aarch32hf) (build 25.152-b76, mixed mode, Evaluation)

~ $  java -XshowSettings 2>&1 | grep sun.arch
    sun.arch.abi = gnueabihf
    sun.arch.data.model = 32
```

means that openJDK isn't set up for hardware float by default? Can this be changed simply by setting this property?

**Comment by wborn** [comment]
On Windows the nrjavaserial library crashes the whole runtime when used with Java 11.

See also https://github.com/NeuronRobotics/nrjavaserial/issues/131 and this [hs_err_pid3924.log](https://github.com/openhab/openhab-distro/files/2674025/hs_err_pid3924.log).

**Comment by BClark09** [comment]
Looks like getting Jython running has a difference in behaviour on JDK11 too:

https://community.openhab.org/t/solved-setting-java-extra-opts-for-jython-changes-logging-location/65531

**Comment by wborn** [comment]
Nowadays the Java 11 Zulu JVMs can be downloaded for most platforms:
* [Linux 64-bit (glibc)](https://www.azul.com/downloads/zulu/zulu-linux/)
* [Linux 64-bit (musl)](https://www.azul.com/downloads/zulu/zulu-download-alpine/) (Alpine)
* [Linux ARM 32/64-bit variants](https://www.azul.com/downloads/zulu-embedded/) (Raspberry Pi's etc.)
* [macOS](https://www.azul.com/downloads/zulu/zulu-mac/)
* [Windows 64-bit](https://www.azul.com/downloads/zulu/zulu-windows/)

**Comment by dbadia** [comment]
@wborn  Regarding 

> On Windows the nrjavaserial library crashes the whole runtime when used with Java 11.

I have a patched version (per NeuronRobotics/nrjavaserial#131) of nrjavaserial-3.14.0.jar which has resolve the issue for me.  Should I submit a PR somewhere?  

All of my prior work has been on the addons so I'm not sure which OH git project the core jars are in.  On my OH runtime, the library is under runtime/system/com/neuronrobotics/nrjavaserial/3.14.0/nrjavaserial-3.14.0.jar

**Comment by wborn** [comment]
I'd prefer all PRs for adding Java 11 support to go to the upstream repository ([NeuronRobotics/nrjavaserial](https://github.com/NeuronRobotics/nrjavaserial)). If you read the issue (https://github.com/NeuronRobotics/nrjavaserial/issues/131) they have the [java11](https://github.com/NeuronRobotics/nrjavaserial/tree/java11) branch for this.

**Comment by J-N-K** [comment]
Unfortunately it looks that there is not much progress over there.

**Comment by wborn** [comment]
That's what's also being said about openHAB. ;-)

**Comment by dbadia** [comment]
Found an issue with Eclipse IDE and Java 11.  I am running Java 11 on my production system (linux) without issue.  UIs work fine.  

In my development environment on Windows/Eclipse, Basic UI works in Java 9, but not in Java 11.  When running with Java 11 I see error pop ups in BasicUI such as `error 404 ProxyServlet /rest/services` and no data (Things, Items, etc) is shown in the UI  

If change the runtime JRE in openHAB_Runtime back to 9 and restart, BasicUI is fully functional again.

Again, this seems specific to Eclipse and/or Windows as my prod setup with JRE 11 on linux works fine.

Anyone else see this issue?  
I can dig deeper but wanted to start by seeing if this is already being tracked

**Comment by davidgraeff** [comment]
Unfortunately `com.eclipsesource.jaxrs` is 4 years old and out of maintenance and the bundled `jersey` is also as old. That is probably not Java 11 compatible.

I'm on my way to replace the entire dependency with the current OSGi approach that doesn't even need a specific library, just the reference implementation. See openhab-core for more information and stay tuned.

**Comment by wborn** [comment]
According to your logging it also fails to allocate a port @dbadia which may also be causing issues : 

```
2019-04-15 19:35:43.190:WARN:oejobis.JettyServerServiceTracker:Start Level: Equinox Container: 1697a2ab-c73a-4746-bb4e-e63903216ca7: 
MultiException[java.net.BindException: Address already in use: bind, java.net.BindException: Address already in use: bind]
```

It looks we might run into SSE issues with the old Jersey version, see also https://github.com/eclipse-ee4j/jersey/issues/3965#issuecomment-440103185. It's easier to stay tuned if you add links (https://github.com/openhab/openhab-core/pull/739) @davidgraeff . :wink:

**Comment by mheidt** [comment]
I am using the latest buster rpi noobs image and installed a 2.5 snapshot.
But OpenHAB doesn't start up:

WARNING: package org.apache.karaf.specs.locator not in java.base
Error starting karaf activator org.apache.karaf.specs.activator.Activator:org/apache/karaf/specs/locator/OsgiLocator

openjdk version "11.0.3" 2019-04-16
OpenJDK Runtime Environment (build 11.0.3+7-post-Raspbian-5)
OpenJDK Server VM (build 11.0.3+7-post-Raspbian-5, mixed mode)

**Comment by wborn** [comment]
> WARNING: package org.apache.karaf.specs.locator not in java.base
Error starting karaf activator org.apache.karaf.specs.activator.Activator:org/apache/karaf/specs/locator/OsgiLocator

I also ran into it and it's fixed for me with the changes in https://github.com/openhab/openhab-distro/pull/991 @mheidt .

**Comment by wborn** [comment]
It seems Gson is also causing warnings due reflection when opening a sitemap in Basic UI:

```
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by com.google.gson.internal.bind.ReflectiveTypeAdapterFactory (file:/home/wouter/software/openhab/instance/userdata/cache/org.eclipse.osgi/21/0/bundleFile) to field java.io.ByteArrayOutputStream.buf
WARNING: Please consider reporting this to the maintainers of com.google.gson.internal.bind.ReflectiveTypeAdapterFactory
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
```

**Comment by wborn** [comment]
Hey @dbadia the UIs work for me in Eclipse with the changes in https://github.com/openhab/openhab-distro/pull/1007.

**Comment by wborn** [comment]
It seems that older MaryTTS versions do work with GA Java 11 releases. In GA releases the `java.version` consists of 3 segments instead of one, i.e. "11.0.5" instead of "11" that was used in the Early Access releases. So I've closed the MaryTTS issues. :slightly_smiling_face:

**Comment by kaikreuzer** [comment]
I think this can be closed now - 3.0 is on Java 11 🎉.

**Comment by tgurr** [comment]
Still shows up as "Pinned issues" on the openhab-distro repository, might want to remove that as well now if not kept on purpose there.

**Comment by kaikreuzer** [comment]
Thanks for noticing, good point!

**Comment by ChrisDumont** [comment]
I'd like some clarification, please. So, 3.0 runs with Java 11. That's good news. But according to the "Installation Overview" openhab2 needs Java 8. I'd like to upgrade my server to Debian Buster which only comes with Java 11. I can, if need be, install openjdk-8-jre from oldstable (Stretch) but I like to avoid mixing OS versions on my servers.

Is there an intention/plan to make openhab2 run with Java 11?

**Comment by wborn** [comment]
Instead of mixing OS versions you can also install the [Zulu Java 8 JVM](https://www.azul.com/downloads/zulu-community/?&version=java-8-lts&os=&os=linux&package=jdk). We do the same with the Debian Buster based openHABian and Docker containers.

There was already a lot of work done to make OH 2.5 compatible with Java 11 so you can try it yourself to see how well it works. I've done some testing with it and the only major issues that I know of are incompatibilities with Jython and the whole JVM crashing on Windows when using add-ons that access serial devices.

When there are add-on specific Java 11 issues it may be possible to fix these in 2.5.x.

**Comment by openhab-5iver** [comment]
> incompatibilities with Jython

Which may be resolved in Jython 2.7.2b3 (I have not tested it with Java 11)

**Comment by J-N-K** [comment]
It will NOT run under Java11 out-of-the-Box. The boot classpath is not working.
